corehq/ex-submodules/casexml/apps/phone/data_providers/case/livequery.py

```diff
@@ -118,12 +118,16 @@ def get_case_hierarchy(domain, cases):
     assert domains == {domain}, "All cases must belong to the same domain"
 
     case_ids = {case.case_id for case in cases}
+    new_cases = get_parent_and_ext_cases(domain, case_ids)
 
+    return cases + new_cases
+
+
+def get_parent_and_ext_cases(domain, case_ids):
```

Comment:
Seems like this change is just to accommodate passing the case IDs instead of the full cases but looking at the callers I think it should be easy enough to pass the full cases. I also think it is necessary because if the app is also using a Data Registry, the cases may not all belong to the same domain and we'd need to segment them. There is already code for doing that, we'd just need to plumb it in:

When creating the search helper we can update the code to return the `DataRegistryHelper` and pass that to the `_RegistryQueryHelper`

```diff
if registry_slug:
-    query_domains = _get_registry_visible_domains(couch_user, domain, case_types, registry_slug)
+    registry_helper = _get_and_validate_registry_helper(couch_user, domain, case_types, registry_slug)
     helper = _RegistryQueryHelper(domain, couch_user, registry_helper)
```

Then in the query helper classes we can add a function:
```python
class _QueryHelper:
    def get_case_hierarchy(self, cases):
        from casexml.apps.phone.data_providers.case.livequery import get_case_hierarchy
        return get_case_hierarchy(self.domain, cases)

class _RegistryQueryHelper:
    def get_case_hierarchy(self, cases):
        return self.registry_helper.get_multi_domain_case_hierarchy(self.couch_user, cases)
```

And then finally use that helper function instead of calling it directly.

Reply:
My main intention of separating it out was actually to be able to grab just the new cases since `get_case_hierarchy` returns both the original and new cases. But `get_parent_and_ext_cases ` (not called `get_parent_host_ext_cases` since commit [da6b314](https://github.com/dimagi/commcare-hq/pull/32587/commits/da6b3145f31536d20533ac734731bc7d07a804f8)) can also take in full cases as a parameter.

Your recommended change makes sense but I'm not sure how the `registry_helper` validation should be handled. For the pseudo `_get_and_validate_registry_helper` function, a `DataRegistryHelper ` is returned if validated, but what should be returned upon a `RegistryNotFound` exception?

Reply:
> My main intention of separating it out was actually to be able to grab just the new cases since `get_case_hierarchy` returns both the original and new cases. But `get_parent_and_ext_cases ` (not called `get_parent_host_ext_cases` since commit [da6b314](https://github.com/dimagi/commcare-hq/pull/32587/commits/da6b3145f31536d20533ac734731bc7d07a804f8)) can also take in full cases as a parameter.

That makes sense. I think the main thing I want to make sure we keep is the domain validation, so if we can move that into the new function as well then 👍🏻 

 
> Your recommended change makes sense but I'm not sure how the `registry_helper` validation should be handled. For the pseudo `_get_and_validate_registry_helper` function, a `DataRegistryHelper ` is returned if validated, but what should be returned upon a `RegistryNotFound` exception?

The error handling can be moved up a level to `get_case_search_results` and if the exception is raised we can just use the normal `_QueryHelper` (the same as if no registry were passed in).